clang/lib/CodeGen/CGHLSLRuntime.cpp

```diff
@@ -280,18 +280,22 @@ void CGHLSLRuntime::annotateHLSLResource(const VarDecl *D, GlobalVariable *GV) {
   const auto *RD = Ty->getAsCXXRecordDecl();
   if (!RD)
     return;
-  const auto *HLSLResAttr = RD->getAttr<HLSLResourceAttr>();
-  const auto *HLSLResClassAttr = RD->getAttr<HLSLResourceClassAttr>();
-  if (!HLSLResAttr || !HLSLResClassAttr)
-    return;
+  // the resource related attributes are on the handle member
+  // inside the record decl
+  for (auto *FD : RD->fields()) {
+    const auto *HLSLResAttr = FD->getAttr<HLSLResourceAttr>();
+    const auto *HLSLResClassAttr = FD->getAttr<HLSLResourceClassAttr>();
+    if (!HLSLResAttr || !HLSLResClassAttr)
+      continue;
 
-  llvm::hlsl::ResourceClass RC = HLSLResClassAttr->getResourceClass();
-  llvm::hlsl::ResourceKind RK = HLSLResAttr->getResourceKind();
-  bool IsROV = HLSLResAttr->getIsROV();
-  llvm::hlsl::ElementType ET = calculateElementType(CGM.getContext(), Ty);
+    llvm::hlsl::ResourceClass RC = HLSLResClassAttr->getResourceClass();
+    llvm::hlsl::ResourceKind RK = HLSLResAttr->getResourceKind();
+    bool IsROV = HLSLResAttr->getIsROV();
+    llvm::hlsl::ElementType ET = calculateElementType(CGM.getContext(), Ty);
 
-  BufferResBinding Binding(D->getAttr<HLSLResourceBindingAttr>());
-  addBufferResourceAnnotation(GV, RC, RK, IsROV, ET, Binding);
+    BufferResBinding Binding(D->getAttr<HLSLResourceBindingAttr>());
+    addBufferResourceAnnotation(GV, RC, RK, IsROV, ET, Binding);
```

Comment:
What if there are >1 fields in the struct that have the resource / resource class attributes on it?  Is it ok to add multiple bindings in that case?

Reply:
It's a great question. With how RWBuffer and  RazterizerOrderedBuffer are implemented, there is exactly one handle, and that handle gets the attributes. I don't believe any future resource object implementations will have multiple handles, or generally, multiple members inside the record decl with the resource attributes.
I am unsure how this interacts with UDTs though. I assume this function is called on decls that have been detected to be resource types, and so this function might be directly called on UDT members that are resources, so there wouldn't be a need to recursively search in the for loop for members within members etc.
@bogner might have an idea for how this works with UDTs?

Reply:
Ultimately, I think all of this is something that we'll allow users to enter into their own source code, so we may need to validate it and be robust against people doing things we don't expect.

Reply:
Given that we're moving to the handle itself really being the resource, I think that if there were multiple fields with the attributes then that really is multiple bindings and the way this is done probably works ok. In any case, this code is going to go away when we implement #95952 anyways, since at that point the information we're currently keeping in `hlsl.uavs`/`hlsl.cbufs` will be carried to the backend by the handle itself rather than in metadata.